library/ssl_tls.c

```diff
@@ -6618,7 +6589,9 @@ int mbedtls_ssl_read( mbedtls_ssl_context *ssl, unsigned char *buf, size_t len )
                 /* With DTLS, drop the packet (probably from last handshake) */
 #if defined(MBEDTLS_SSL_PROTO_DTLS)
                 if( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM )
-                    return( MBEDTLS_ERR_SSL_WANT_READ );
+                {
+                    continue;
+                }
```

Comment:
@RonEld The previous meaning of `WANT_READ` at this point was to signal to the user that he should call `mbedtls_ssl_read` again for further message processing. This is not what `WANT_READ` is supposed to express, and now we immediately do the fetching and processing of the next message within `mbedtls_ssl_read`. 

So in short, we're dropping / ignoring the packet in both cases, but in a slightly different way. 

The most direct way to functionally mimic a user calling `mbedtls_ssl_read` immediately again would be to do a recursive call or to jump to the beginning of the function. A recursive call is no option because we don't rely on tail-recursive optimizations by the compiler, and a goto jump to the beginning isn't good style. It seemed to me that a reasonble place to jump to for the next iteration would be the `if( ssl->in_offt == NULL)` check, but it needs to be double-checked that the part from the beginning of `mbedtls_ssl_read` skipped this way is indeed irrelevant for subsequent iterations.